File: app/src/main/java/com/fenfei/myapplicationdemo/ui/ListViewActivity.java
package com.fenfei.myapplicationdemo.ui;

import android.content.Context;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ListView;

import com.fenfei.myapplicationdemo.R;
import com.fenfei.myapplicationdemo.designpattern.factroypattern.ChicagoPizzaStore;
import com.fenfei.myapplicationdemo.designpattern.factroypattern.PizzaStrore;

import java.util.LinkedList;
import java.util.List;

public class ListViewActivity extends AppCompatActivity {

    private ListView mListView_;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_list_view);

        mListView_ = (ListView) findViewById(R.id.listview);

        final LinkedList<String> dataset = new LinkedList<>();

        String[] arr = new String[]{"1", "2", "3", "4", "5", "6", "7", "8", "9",
                "10", "11", "12", "13", "14", "15", "16",
                "17", "18", "19", "20"};
        for (String s : arr) {
            dataset.add(s);
        }
        final MyAdapter adapter = new MyAdapter(this, android.R.layout.simple_list_item_1, dataset);
        mListView_.setAdapter(adapter);

        mListView_.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                String str = dataset.get(position);
                dataset.remove(position);
                adapter.notifyDataSetChanged();
                mListView_.setSelection(position);
                Log.e("listview", "onItemClick: " + str);

                PizzaStrore strore = new ChicagoPizzaStore();
                strore.orderPizza("Chicago");
            }
        });
    }

    private class MyAdapter extends ArrayAdapter<String> {

        public MyAdapter(Context context, int resource, List<String> objects) {
            super(context, resource, objects);
        }

    }

}

File: camera/src/main/java/com/fenfei/camera/kotlins/ListViewActivity.kt
package com.fenfei.camera.kotlins

import android.content.Context
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ListView
import android.widget.TextView
import android.widget.Toast
import com.fenfei.camera.R

/**
 * 测试Kotiln语法下的listview
 */
class ListViewActivity : AppCompatActivity() {

    var resId: Int = R.layout.item_kt_list

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_list_view)

        var lists = arrayListOf("a", "b", "a", "b", "a", "b")
        var user = User("shefenfei", "")
        println(user.toString())

        var listview = findViewById(R.id.listview) as ListView
        var adapter = MyAdapter(this, resId, lists)
        listview.adapter = adapter

        listview.setOnItemClickListener { adapterView, view, position, listener -> showSomething(lists[position]) }
//        listview.setOnClickListener { showSomething(lists[0]) }
//        listview.setOnClickListener { view -> showSomething(lists[0]) }

        testLambda()

    }

    private fun testLambda() {
//        val datas = listOf(User("" ,"shefenfei" , 1))
    }


    var callback: (() -> Unit)? = {
        Log.e("TAG" , ".........")
    }

    var callback1: ((str: String) -> Unit)? = {
        Log.e("TAG", "")
    }


    //show
    fun showSomething(text: String) {
        Toast.makeText(this, text, Toast.LENGTH_SHORT).show()
        callback?.invoke()
        callback1?.invoke("aaa")
    }

    class MyAdapter(context: Context?, resource: Int, objects: List<String>) : ArrayAdapter<String>(context, resource, objects) {

        var datasets: List<String> = emptyList<String>()
        var res: Int = 0
        var c: Context? = null

        init {
            c = context
            datasets = objects
            res = resource
        }

        override fun getView(position: Int, convertView: View?, parent: ViewGroup?): View {
            var view = LayoutInflater.from(c).inflate(res, parent, false)
            var tv_name = view.findViewById(R.id.item_name) as TextView
            tv_name.text = datasets[position]
            return view
        }

        override fun getCount(): Int {
            return datasets.size
        }
    }
}
